llms.txt

```diff
@@ -11289,6 +11290,780 @@ Doc-Content: https://docs.kluster.ai/tutorials/klusterai-api/image-analysis.ipyn
+    "            \n",
+    "        tool_results.append({\n",
+    "            \"tool_call_id\": tool_call.id,\n",
+    "            \"function_name\": function_name,\n",
+    "            \"result\": result\n",
+    "        })\n",
+    "    \n",
+    "    # Create a new message with the tool results\n",
+    "    messages = [\n",
+    "        {\"role\": \"user\", \"content\": original_query},\n",
+    "        message.model_dump(),\n",
+    "    ]\n",
+    "    \n",
+    "    # Add the tool results\n",
+    "    for result in tool_results:\n",
+    "        messages.append({\n",
+    "            \"role\": \"tool\",\n",
+    "            \"tool_call_id\": result[\"tool_call_id\"],\n",
+    "            \"content\": json.dumps(result[\"result\"])\n",
+    "        })\n",
+    "    \n",
+    "    # Get the final response\n",
+    "    final_response = client.chat.completions.create(\n",
+    "        model=\"klusterai/Meta-Llama-3.1-8B-Instruct-Turbo\",\n",
+    "        messages=messages\n",
+    "    )\n",
+    "    \n",
+    "    return final_response.choices[0].message.content\n",
+    "\n",
+    "# Test with a complex query that might use multiple tools\n",
+    "multi_query = \"If the Earth's temperature rises by 2 degrees, what percentage increase is that from the current average global temperature of 15 degrees Celsius?\"\n",
+    "multi_response = run_with_tools(multi_query, multi_tools)\n",
+    "final_multi_result = process_multi_tool_calls(multi_response, multi_query)\n",
+    "\n",
+    "print(final_multi_result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Real-world Use Case: Document Analysis with Tools\n",
+    "\n",
+    "Here we treat an entire report as context and give the model two powers: read the text and, if it spots numbers that need crunching, invoke the calculator. The helper wraps the document and question into a single prompt, passes in our multi_tools array, and lets the model decide the workflow. In the revenue example the LLM extracts $1.2 M → $1.8 M, calls the calculator to compute the delta, then replies in prose: “The percentage increase in revenue was 50 %.” This pattern scales to meeting minutes, legal contracts, or log files- any place the model must combine natural language comprehension with deterministic math."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "According to the report, the percentage increase in revenue was 50%.\n"
+     ]
+    }
+   ],
+   "source": [
+    "def document_analysis_with_tools(document, question):\n",
+    "    # Prepare the prompt\n",
+    "    prompt = f\"\"\"\n",
+    "Document: \n",
+    "{document}\n",
+    "\n",
+    "Question about the document: {question}\n",
+    "\n",
+    "Please answer the question based on the document. If calculations are needed, use the calculator tool.\n",
```

Comment:
Format this python code,. it is not properly formatted. I tend to use -> https://codebeautify.org/python-formatter-beautifier